proposals/service_api_calls.md

```diff
@@ -0,0 +1,186 @@
+# Meta
+[meta]: #meta
+- Name: Service API Call
+- Start Date: December 20, 2022
+- Update data (optional):  December 20, 2022
+- Author(s): @JimBugwadia
+
+
+# Table of Contents
+[table-of-contents]: #table-of-contents
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+- [CRD Changes (OPTIONAL)](#crd-changes-optional)
+
+# Overview
+[overview]: #overview
+
+Support HTTP/S GET and POST calls to JSON web services.
+
+# Definitions
+[definitions]: #definitions
+
+
+# Motivation
+[motivation]: #motivation
+
+Kyverno versions 1.9 and earlier support API calls to the Kubernetes API server and loading data from ConfigMap resources. 
+
+However, policies often require data from other services e.g. a metrics service. Currently, this requires writing custom code to periodically fetch data from the services and store the results in a ConfigMap or other Kubernetes resource that can be retrieved by Kyverno. This feature allows the service to be directly queried, eliminating the need to write custom services and create Kubernetes custom resources for policy data.
+
+In other cases, some policies may require delegation of processing to other controllers. For example, a separate controller may periodically fetch and cache external data which is used to process policies. With this feature, Kyverno can send the service a POST request with context data, and the service can respond back with results that are used to make the policy decision. 
+
+Finally, in some scenarios 3rd party library code or binaries need to be executed to make a policy decision. Currently, the only way to support these "extensions" is to directly update the Kyverno core. With this feature, it will be possible to write Kyverno extensions as separate services hence decoupling the Kyverno code from 3rd party libraries or binaries.
+
+# Proposal
+
+This feature enables:
+- HTTP GET and POST API calls to any service
+- Encryption and server authentication using a certificate bundle
+- Client authentication using Service Account token volume projection 
```

Comment:
@JimBugwadia I think we should also add client authentication using the `Authorization` header. Most of the web services and k8s itself use JWTs for authentication. Having it available as a way to authenticate would dramatically help with the usability of this feature.

Reply:
In general I believe we should support the following methods of authentication -

- Basic
- JWT
- certs/mTLS

The above 3 are also supported by kubernetes.

Reply:
Example of how goreleaser implements this https://goreleaser.com/customization/upload/#username

Reply:
@samj1912 - yes, the intent was to add an API token to the HTTP Authorization header. I've clarified.